### ISSUE TITLE
Expose simulated throttling specific requestIdleCallback params

### ISSUE THREAD

**Issue by LukasAuerMstage** [GoogleChrome/lighthouse#10693]
#### Description
As i was consistently experiencing mediocre TTI Scores in my Applications i started splitting Code and applied the Idle-until-urgent Pattern (using the idlize library Philip Walton, which is using requestIdleCallback under the hood) quite agressively until i had no visible Long Tasks in my Performance Tab. However even after making these optimizations i still experienced TTI Scores above 4s while having FMP Values of about 1.5s. 

So i started making some isolated Tests and came to the conclusion that Lighthouse seems to find Long Tasks that are not reflected in the Performance Tab as such. In these isolated Tests this is obviously not a big deal as it is a small HTML Document with a single Script Tag, but in a larger Application this leads to bad TTI Scores which are harming the overall Score and there is no real way to identify and debug them, as they cannot be detected in the Performance Tab.

#### Provide the steps to reproduce
1. Clone Repository (https://github.com/LukasAuerMstage/lighthouse-tti)
2. Run Lighthouse Audit against the test HTML Page

#### What is the current behavior?
TTI ist higher than FMP although there are no visible Long Tasks in the Performance Tab. This can be seen in the Lighthouse Result: (FMP 0.6, TTI 0.9)
https://github.com/LukasAuerMstage/lighthouse-tti/blob/master/lh-score.PNG
Task Timing can be seen in this Screenshot:
https://github.com/LukasAuerMstage/lighthouse-tti/blob/master/performance-tab.png
or the Performance Profile:
https://github.com/LukasAuerMstage/lighthouse-tti/blob/master/Profile-20200502T162332.json

Note: Profile was created with 6 Times CPU Throttling and Slow 3g.

#### What is the expected behavior?
According to the Specification TTI should be the same as FMP (or DomContentLoaded respectively, whatever fires last) if no Long Tasks were found. As Tasks are split into Chunks that do not exceed the 50ms Limit (about 30ms each), this should be the case.

#### Further Explanation
This Test Document obviously takes a very naive approach with just scheduling 2 Timeouts while simulating a Sleep for 30ms each but as i stated above i also experienced similar Results in more Complex Environments while structuring similar Chunks of Execution Time in requestIdleCallbacks.

According to my Experience every Chunk of Code that appears to be longer than 20-30ms (while still remaining below 50ms) in the Chrome Performance Tab potentially produces a Long Task under the Hood and therefore ruins the TTI Score.

So as long as i only have 10ms Chunks and i run them sequentially, i can process as many Scripts as i want without risking to produce an unseen Long Task. But as soon as one Chunk somewhere takes about 30ms, my TTI Score gets in the red.

#### Debugging the issue
I looked a bit into the Lighthouse Code and ran some local Audits while printing out Long Tasks in my console and indeed Lighthouse finds a Long Tasks that is responsible for the Scoring:

```
node: CPUNode {
    _id: '12440.65069837946',
    _isMainDocument: false,
    _dependents: [],
    _dependencies: [ [CPUNode], [NetworkNode] ],
    _event: {
      args: {},
      cat: 'disabled-by-default-devtools.timeline',
      dur: 30939,
      name: 'RunTask',
      ph: 'X',
      pid: 17420,
      tdur: 30828,
      tid: 12440,
      ts: 65069837946,
      tts: 648811
    },
    _childEvents: [ [Object], [Object], [Object], [Object], [Object], [Object] ]
  },
  endTime: 848.281,
  timing: { startTime: 724.281, endTime: 848.281, duration: 124 }
```

When printing the Child Events, the output looks as such:

```
TimerFire
ScheduledAction::execute
v8.callFunction
FunctionCall
V8.Execute
UpdateCounters
```

#### Specification (TTI)
Find a the first 5 second window W after FMP such that 
W overlaps no tasks longer than 50ms
For all timestamp t in W, number of resource requests in flight at t is no more than 2. 

Now find the last long task L before W. 
Consistently Interactive Candidate is the end of L
In the case there is no long task before L, Consistently Interactive Candidate = FMP.

Now, `Consistently Interactive = max(Consistently Interactive Candidate, DOMContentLoadedEventEnd)`

#### Environment Information
* Affected Channels: Lighthouse Devtools, as well as Lighthouse called programmatically and Lighthouse CLI (overall, its reproduceable in all available Lighthouse Channels)
* Lighthouse version: 5.7.0
* Chrome version: Version 81.0.4044.129 (64-Bit)
* Node.js version: v12.16.1
* Operating System: Windows for this Performance Test but my Sites, which experience similar Results, run on Linux Servers

#### Conclusion
I dont know if these are false Long Tasks (i assume they are not) and this is a Bug in the end, but at least this Behaviour should be reflected in the Performance Tab. As of right now it is very hard to optimize for a good TTI Score in my opinion. As Lighthouse 5.x has a very strong emphasis on TTI Scrores this is actually quite problematic. I would be happy if you would look into the matter and tell me if this is actually expected Behaviour, a bug or if i am missing something entirely. Thank!

**Comment by patrickhulce** [comment]
Thanks for filing such a detailed issue @LukasAuerMstage! 

Unfortunately, the tasks created by idilize and that example repo are crafted in a way that defeats simulated throttling (the Lighthouse default). Simulated throttling looks at how long tasks take to execute on a powerful device and uses that to estimate how long they would have taken on a weaker device. The example repo creates tasks that take 30ms no matter how slow the device is. Simulated throttling sadly does not know whether the microtasks had an intelligent queuing mechanism like idilize *and* that the chunks were small enough  to not overextend that threshold on their own. 

I would not expect [applied throttling methods](https://github.com/GoogleChrome/lighthouse/blob/master/docs/throttling.md) to suffer from this problem, so the workaround would be using `--throttling-method=devtools|provided`

**Comment by LukasAuerMstage** [comment]
@patrickhulce Thanks for the quick reply, the Flag will definetely help a lot for my Testing and optimizing my sites for a good user experience.

Besides that i would be curious how this affects overall scoring on public Lighthouse Channels, like PageSpeedInsights and web.dev.

As idilize uses requestIdleCallback and within that it tries to process as many small tasks as possible within the remaining time of the deadline, there will always be a case that simulated throttling will detect ist as a long task, is that correct? If the complete deadline Time of 50ms is used in one Callback, simulated throttling would probably behave similarly as in my test.

I assume that PageSpeedInsights for example uses the default Lighthouse throttling and therefore the approach of queueing up Tasks as idilize does, paired with the fact that simulated throttling is in place, can be a potential showstopper for these ratings if I understood it correctly.

**Comment by patrickhulce** [comment]
> there will always be a case in which simulated throttling will detect it as a long task, is that correct?

When the tasks queued for idlize take more than 12.5ms on the test device, then they will appear as long tasks under 4x simulated throttling, correct. 

> I assume that PageSpeedInsights for example uses the default Lighthouse throttling

Correct, PSI uses simulated throttling (and web.dev uses the same backend as PSI).


IMO, the workaround here is that the threshold of 50ms in the polyfill for rIC should be adjusted to ~12ms under simulated throttling. Just because idlize is being used doesn't mean the tasks it's generating aren't long tasks on slow devices if each unit of work in the idle callback is large.

Maybe there's something Lighthouse could to do shim rIC or expose the simulation constant on `window` so that pages could cooperate with it? With a 12ms threshold instead of 50ms the developer is still warned of legitimate long task issues when using the library and won't be penalized for doing the sophisticated thing.

**Comment by LukasAuerMstage** [comment]
As much as i would love a window variable like this I would assume it could open up cases for abuse. I dont know if thats a concerne of yours, but if developers could interact with this constant at least server rendered pages could strip out JS in its entirety once this variable is present and therefore get almost perfect LH Results, which they could present to their clients, that dont reflect the acutal user experience.

**Comment by patrickhulce** [comment]
> I dont know if thats a concerne of yours, but if developers could interact with this constant at least server rendered pages could strip out JS in its entirety once this variable is present and therefore get almost perfect LH Results, which they could present to their clients, that dont reflect the acutal user experience.

While I am concerned about this general problem, we are already forced to announce requests from PSI in the UA, so shimming one API in a difficultish to detect way is not a large concern IMO.

I agree with your point about avoiding any sort of "this version of the page for LH" effect the `window` property might have which is why my preferred route would be the rIC shim to have it work seamlessly with `deadline.timeRemaining()` based approaches.

**Comment by LukasAuerMstage** [comment]
https://github.com/LukasAuerMstage/lighthouse/commit/46eddb7f21bbeefce54efa20fc4cfbc5a77cd6e0

Something like that could work, contrary to existing rIC shims out there, it would still use the native rIC method under the hood instead of setTimeout.

**Comment by LukasAuerMstage** [comment]
Looking at it, the timeout passing is obviously wrong, should look more like this
https://github.com/LukasAuerMstage/lighthouse/commit/4de0fed1c8854e1903ac80da0971ea230a5c0b22

**Comment by patrickhulce** [comment]
Yeah basically something just like that! :) WDYT?

**Comment by LukasAuerMstage** [comment]
Yeah i think so too. For a second i thought about a possibility to extend the IdleDeadline directly but that wouldn't work and this approach is probably cleaner anyway.

**Comment by LukasAuerMstage** [comment]
@patrickhulce As i havent done any LH contributions before i am not sure if that is a valid issue that, but if so can i still start an PR? And if so, should i write tests for the functionality?

**Comment by patrickhulce** [comment]
That would be great @LukasAuerMstage! 

I think the only substantial change from what you've done already is that we'll only want to wrap this when `throttlingMethod=simulate`. The way we'll want to test this is via what we call "smoke tests".

There are 3 parts to a smoke test:

1. The fixture, example: https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-cli/test/fixtures/delayed-fcp.html. We'll create a sample page that uses requestIdleCallback in a loop that executes in 1ms chunks until deadline is hit (this should pass without TTI penalty) and another that executes in 50ms chunks (this should fail). You can use a query param instead of creating separate pages so everything is colocated.
2. The expectation, add our assertions for the two test pages to https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-cli/test/smokehouse/test-definitions/lantern/lantern-expectations.js
3. The test definition, if you followed the steps of 2 you don't need to do anything here. But this is where the magic is happening.
https://github.com/GoogleChrome/lighthouse/blob/4c0a76b8716fda81457335a75f7e523903eb4d20/lighthouse-cli/test/smokehouse/test-definitions/core-tests.js#L62-L64

**Comment by LukasAuerMstage** [comment]
@patrickhulce great, thanks for the clarification. I will submit the PR once i have finished the smoke tests. Regarding the throttling method, i have actually included that already in my first commit:
https://github.com/LukasAuerMstage/lighthouse/commit/46eddb7f21bbeefce54efa20fc4cfbc5a77cd6e0

`if (settings.throttlingMethod === 'simulate') {`

**Comment by patrickhulce** [comment]
Oh I missed that I only looked at the second 😆 excellent then even less work for you! :)

**Comment by patrickhulce** [comment]
Hey @LukasAuerMstage just checking in on this to see if you were still planning on opening a PR :) No rush or anything just looking for status

**Comment by LukasAuerMstage** [comment]
@patrickhulce Yeah i definately will submit it, sorry for the delay, probably within the next couple of days.

**Comment by patrickhulce** [comment]
I think we've fully addressed this with #11032 🎉  thanks for your contribution @LukasAuerMstage!
